kernel/mmanutils.c

```diff
@@ -517,219 +571,288 @@ int myst_get_free_ram(size_t* size)
-                    p->offset,
-                    p->pathname);
-
-                if (n >= (int)sizeof(locals->maps_entry))
-                    ERAISE(-ERANGE);
-
-                // Insert new entry at beginning
-                ECHECK(myst_buf_insert(
-                    vbuf, 0, locals->maps_entry, strlen(locals->maps_entry)));
+                if (v.pids[i] == (uint32_t)pid)
+                {
+                    size_t n = 1;
+                    size_t len;
+                    int fd = v.fdmappings[i].fd;
+
+                    uint32_t used = v.fdmappings[i].used;
+                    uint64_t offset = v.fdmappings[i].offset;
+                    myst_refstr_t* pathname = v.fdmappings[i].pathname;
+                    int prot = 0;
+                    bool consistent = false;
+                    char* str;
+                    int flags = 0;
+
+                    if (myst_mman_get_prot(
+                            &_mman, addr, PAGE_SIZE, &prot, &consistent) != 0)
+                    {
+                        assert("myst_mman_get_prot() failed\n");
+                    }
+
+                    /* count consecutive pages with same traits */
+                    for (size_t j = i + 1; j < index + count; j++)
+                    {
+                        int tmp_prot = 0;
+
+                        /* if the pid changes */
+                        if (v.pids[j] != (uint32_t)pid)
+                            break;
+
+                        if (v.fdmappings[i].used != used)
+                            break;
+
+                        /* if the fd changes */
+                        if (v.fdmappings[j].fd != fd)
+                            break;
+
+                        if (myst_mman_get_prot(
+                                &_mman,
+                                addr + (n * PAGE_SIZE),
+                                PAGE_SIZE,
+                                &tmp_prot,
+                                &consistent) != 0)
+                        {
+                            assert("myst_mman_get_prot() failed\n");
+                        }
+
+                        if (tmp_prot != prot)
+                            break;
+
+                        n++;
+                    }
+
+                    len = n * PAGE_SIZE;
+
+                    if (!used)
+                        fd = -1;
+
+                    /* format the output */
+                    if (_format_proc_maps_entry(
+                            addr,
+                            len,
+                            prot,
+                            flags,
+                            offset,
+                            (pathname ? pathname->data : ""),
+                            &str) == 0)
+                    {
+                        ECHECK(myst_buf_insert(vbuf, 0, str, strlen(str)));
+                        free(str);
+                    }
+
+                    i += n;
+                    addr += len;
+                    length -= len;
```

Comment:
`length` is unused in the critical section.

Reply:
Good find. I removed it.